From 42980fe282cb09ee2259b200f290bf239090647f Mon Sep 17 00:00:00 2001
From: David Heidelberg <david@ixit.cz>
Date: Mon, 1 Jun 2026 00:20:07 +0500
Subject: [PATCH] drm/panel: samsung: Add driver for Samsung EA8074
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Add a DRM panel driver for the Samsung EA8074 display driver IC.
This DDIC and asscoaited panel is found in:
 - Comma.AI box
 - Xiaomi Mi 8

Based on work of:
 - Eralp Çelebi
 - Max Furman <taygoth@gmail.com>

Signed-off-by: David Heidelberg <david@ixit.cz>
---
 .../display/panel/samsung,ea8074.yaml         |  78 ++++
 drivers/gpu/drm/panel/Kconfig                 |  15 +
 drivers/gpu/drm/panel/Makefile                |   1 +
 drivers/gpu/drm/panel/panel-samsung-ea8074.c  | 422 ++++++++++++++++++
 4 files changed, 516 insertions(+)
 create mode 100644 Documentation/devicetree/bindings/display/panel/samsung,ea8074.yaml
 create mode 100644 drivers/gpu/drm/panel/panel-samsung-ea8074.c

diff --git a/Documentation/devicetree/bindings/display/panel/samsung,ea8074.yaml b/Documentation/devicetree/bindings/display/panel/samsung,ea8074.yaml
new file mode 100644
index 000000000000..c2e8c246bd95
--- /dev/null
+++ b/Documentation/devicetree/bindings/display/panel/samsung,ea8074.yaml
@@ -0,0 +1,78 @@
+# SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
+%YAML 1.2
+---
+$id: http://devicetree.org/schemas/display/panel/samsung,ea8074.yaml#
+$schema: http://devicetree.org/meta-schemas/core.yaml#
+
+title: Samsung EA8074 AMOLED DDIC
+
+maintainers:
+  - FIXME
+
+description: |
+  The Samsung EA8074 is display driver IC with connect panel.
+
+allOf:
+  - $ref: panel-common.yaml#
+
+properties:
+  compatible:
+    items:
+      - enum:
+           # Samsung 6.01 inch, 1080x2160 pixels
+          - samsung,ea8074-ams601nt01
+           # Samsung 6.21 inch, 1080x2248 pixels
+          - samsung,ea8074-ams621rc
+           # Samsung 5.9 inch, 1080x2160 pixels
+          - samsung,ea8074-l590nt06
+      - const: samsung,ea8074
+
+  reg:
+    maxItems: 1
+
+  elvdd-supply:
+    description: EL driving positive voltage supply
+
+  elvss-supply:
+    description: EL driving negative voltage supply
+
+  vddio-supply:
+    description: I/O voltage supply
+
+required:
+  - compatible
+  - reg
+  - reset-gpios
+  - elvdd-supply
+  - elvss-supply
+  - vddio-supply
+
+additionalProperties: false
+
+examples:
+  - |
+    #include <dt-bindings/gpio/gpio.h>
+
+    dsi {
+        #address-cells = <1>;
+        #size-cells = <0>;
+
+        panel@0 {
+            compatible = "samsung,ea8074-ams621rc", "samsung,ea8074";
+            reg = <0>;
+
+            reset-gpios = <&tlmm 6 GPIO_ACTIVE_LOW>;
+
+            vddio-supply = <&vreg_l14a_1p8>;
+            elvdd-supply = <&lab>;
+            elvss-supply = <&ibb>;
+
+            port {
+                panel_in: endpoint {
+                    remote-endpoint = <&dsi0_out>;
+                };
+            };
+        };
+    };
+
+...
diff --git a/drivers/gpu/drm/panel/Kconfig b/drivers/gpu/drm/panel/Kconfig
index 407c5f6a268b..386406c2df81 100644
--- a/drivers/gpu/drm/panel/Kconfig
+++ b/drivers/gpu/drm/panel/Kconfig
@@ -762,6 +762,21 @@ config DRM_PANEL_SAMSUNG_DB7430
 	  DB7430 DPI display controller used in such devices as the
 	  LMS397KF04 480x800 DPI panel.
 
+config DRM_PANEL_SAMSUNG_EA8074
+	tristate "Samsung EA8074 DSI command mode panel"
+	depends on OF
+	depends on DRM_MIPI_DSI
+	depends on BACKLIGHT_CLASS_DEVICE
+	help
+	  Say Y or M here if you want to enable support for the
+	  Samsung EA8074 display driver IC and connected panel.
+
+	  Currently supported panels:
+
+	    Samsung AMS601NT01 (found in the Huawei Mate 10 Pro)
+	    Samsung AMS621RC (found in the Xiaomi Mi 8 smartphone)
+	    Samsung L590NT06 (found in the Comma three)
+
 config DRM_PANEL_SAMSUNG_LD9040
 	tristate "Samsung LD9040 RGB/SPI panel"
 	depends on OF && SPI
diff --git a/drivers/gpu/drm/panel/Makefile b/drivers/gpu/drm/panel/Makefile
index 3615a761b44f..9989a96f8dea 100644
--- a/drivers/gpu/drm/panel/Makefile
+++ b/drivers/gpu/drm/panel/Makefile
@@ -74,6 +74,7 @@ obj-$(CONFIG_DRM_PANEL_SAMSUNG_AMS581VF01) += panel-samsung-ams581vf01.o
 obj-$(CONFIG_DRM_PANEL_SAMSUNG_AMS639RQ08) += panel-samsung-ams639rq08.o
 obj-$(CONFIG_DRM_PANEL_SAMSUNG_ATNA33XC20) += panel-samsung-atna33xc20.o
 obj-$(CONFIG_DRM_PANEL_SAMSUNG_DB7430) += panel-samsung-db7430.o
+obj-$(CONFIG_DRM_PANEL_SAMSUNG_EA8074) += panel-samsung-ea8074.o
 obj-$(CONFIG_DRM_PANEL_SAMSUNG_LD9040) += panel-samsung-ld9040.o
 obj-$(CONFIG_DRM_PANEL_SAMSUNG_S6D16D0) += panel-samsung-s6d16d0.o
 obj-$(CONFIG_DRM_PANEL_SAMSUNG_S6D27A1) += panel-samsung-s6d27a1.o
diff --git a/drivers/gpu/drm/panel/panel-samsung-ea8074.c b/drivers/gpu/drm/panel/panel-samsung-ea8074.c
new file mode 100644
index 000000000000..192642332bda
--- /dev/null
+++ b/drivers/gpu/drm/panel/panel-samsung-ea8074.c
@@ -0,0 +1,422 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * DRM driver for the Samsung EA8074 display driver IC and associated panel.
+ *
+ * Copyright 2022 Eralp Çelebi <eralp.celebi.personal@gmail.com>
+ * Copyright 2026 Max Furman <taygoth@gmail.com>
+ * Copyright 2026 David Heidelberg <david@ixit.cz>
+ */
+
+#include <linux/backlight.h>
+#include <linux/delay.h>
+#include <linux/gpio/consumer.h>
+#include <linux/module.h>
+#include <linux/of.h>
+#include <linux/regulator/consumer.h>
+
+#include <video/mipi_display.h>
+
+#include <drm/drm_mipi_dsi.h>
+#include <drm/drm_modes.h>
+#include <drm/drm_panel.h>
+#include <drm/drm_probe_helper.h>
+
+struct samsung_ea8074_panel_desc {
+	const struct drm_display_mode *mode;
+	int (*init_sequence)(struct mipi_dsi_multi_context *dsi_ctx);
+};
+
+struct samsung_ea8074 {
+	struct drm_panel panel;
+	struct mipi_dsi_device *dsi;
+	struct regulator_bulk_data *supplies;
+	struct gpio_desc *reset_gpio;
+	const struct samsung_ea8074_panel_desc *desc;
+};
+
+static const struct regulator_bulk_data ea8074_supplies[] = {
+	{ .supply = "vddio",
+	  .init_load_uA = 62000 },
+	{ .supply = "elvdd",
+	  .init_load_uA = 100000 },
+	{ .supply = "elvss",
+	  .init_load_uA = 100000 },
+};
+
+static inline
+struct samsung_ea8074 *to_samsung_ea8074(struct drm_panel *panel)
+{
+	return container_of(panel, struct samsung_ea8074, panel);
+}
+
+#define ea8074_test_key_on_lvl2(ctx) \
+	mipi_dsi_dcs_write_seq_multi(ctx, 0xf0, 0x5a, 0x5a)
+#define ea8074_test_key_off_lvl2(ctx) \
+	mipi_dsi_dcs_write_seq_multi(ctx, 0xf0, 0xa5, 0xa5)
+#define ea8074_test_key_on_lvl3(ctx) \
+	mipi_dsi_dcs_write_seq_multi(ctx, 0xfc, 0x5a, 0x5a)
+#define ea8074_test_key_off_lvl3(ctx) \
+	mipi_dsi_dcs_write_seq_multi(ctx, 0xfc, 0xa5, 0xa5)
+
+static void samsung_ea8074_reset(struct samsung_ea8074 *ctx)
+{
+	gpiod_set_value_cansleep(ctx->reset_gpio, 0);
+	usleep_range(1000, 2000);
+	gpiod_set_value_cansleep(ctx->reset_gpio, 1);
+	usleep_range(1000, 2000);
+	gpiod_set_value_cansleep(ctx->reset_gpio, 0);
+	usleep_range(10000, 11000);
+}
+
+static int l590nt06_init_sequence(struct mipi_dsi_multi_context *dsi_ctx)
+{
+	mipi_dsi_dcs_exit_sleep_mode_multi(dsi_ctx);
+	mipi_dsi_usleep_range(dsi_ctx, 10000, 11000);
+
+	mipi_dsi_dcs_set_column_address_multi(dsi_ctx, 0x0000, 0x0437);
+	mipi_dsi_usleep_range(dsi_ctx, 10000, 11000);
+
+	mipi_dsi_dcs_set_page_address_multi(dsi_ctx, 0x0000, 0x0800 | 0x006f);
+	mipi_dsi_usleep_range(dsi_ctx, 10000, 11000);
+
+	mipi_dsi_dcs_set_tear_on_multi(dsi_ctx, MIPI_DSI_DCS_TEAR_MODE_VBLANK);
+	mipi_dsi_usleep_range(dsi_ctx, 10000, 11000);
+
+	mipi_dsi_dcs_write_seq_multi(dsi_ctx, MIPI_DCS_WRITE_CONTROL_DISPLAY,
+				     0x20);
+	mipi_dsi_usleep_range(dsi_ctx, 10000, 11000);
+
+	mipi_dsi_dcs_write_seq_multi(dsi_ctx, MIPI_DCS_SET_DISPLAY_BRIGHTNESS,
+				     0x00, 0x00);
+	mipi_dsi_usleep_range(dsi_ctx, 10000, 11000);
+
+	mipi_dsi_dcs_write_seq_multi(dsi_ctx, MIPI_DCS_WRITE_POWER_SAVE, 0x00);
+	mipi_dsi_usleep_range(dsi_ctx, 10000, 11000);
+
+	/* Unlock Samsung vendor registers */
+	ea8074_test_key_on_lvl2(dsi_ctx);
+	mipi_dsi_usleep_range(dsi_ctx, 10000, 11000);
+	mipi_dsi_dcs_write_seq_multi(dsi_ctx, 0xb0, 0x06);
+	mipi_dsi_usleep_range(dsi_ctx, 10000, 11000);
+	mipi_dsi_dcs_write_seq_multi(dsi_ctx, 0xef, 0x35);
+	mipi_dsi_usleep_range(dsi_ctx, 10000, 11000);
+	mipi_dsi_dcs_write_seq_multi(dsi_ctx, 0xcc, 0x55, 0x12);
+	mipi_dsi_usleep_range(dsi_ctx, 10000, 11000);
+
+	ea8074_test_key_on_lvl3(dsi_ctx);
+	mipi_dsi_usleep_range(dsi_ctx, 10000, 11000);
+	mipi_dsi_dcs_write_seq_multi(dsi_ctx, 0xb0, 0x01);
+	mipi_dsi_usleep_range(dsi_ctx, 10000, 11000);
+	mipi_dsi_dcs_write_seq_multi(dsi_ctx, 0xd2, 0x20);
+	mipi_dsi_usleep_range(dsi_ctx, 10000, 11000);
+	mipi_dsi_dcs_write_seq_multi(dsi_ctx, 0xb0, 0x05);
+	mipi_dsi_usleep_range(dsi_ctx, 10000, 11000);
+	mipi_dsi_dcs_write_seq_multi(dsi_ctx, 0xd2, 0x40);
+	mipi_dsi_usleep_range(dsi_ctx, 10000, 11000);
+	ea8074_test_key_off_lvl3(dsi_ctx);
+	mipi_dsi_usleep_range(dsi_ctx, 10000, 11000);
+
+	ea8074_test_key_off_lvl2(dsi_ctx);
+	mipi_dsi_msleep(dsi_ctx, 110);
+
+	mipi_dsi_dcs_write_seq_multi(dsi_ctx, MIPI_DCS_WRITE_MEMORY_START);
+	mipi_dsi_usleep_range(dsi_ctx, 10000, 11000);
+
+	mipi_dsi_dcs_set_display_on_multi(dsi_ctx);
+	mipi_dsi_usleep_range(dsi_ctx, 10000, 11000);
+
+	/* Post display-on vendor configuration */
+	ea8074_test_key_on_lvl2(dsi_ctx);
+	mipi_dsi_usleep_range(dsi_ctx, 10000, 11000);
+	mipi_dsi_dcs_write_seq_multi(dsi_ctx, 0xb0, 0x05);
+	mipi_dsi_usleep_range(dsi_ctx, 10000, 11000);
+	mipi_dsi_dcs_write_seq_multi(dsi_ctx, 0xb1, 0x40);
+	mipi_dsi_usleep_range(dsi_ctx, 10000, 11000);
+	mipi_dsi_dcs_write_seq_multi(dsi_ctx, 0xb0, 0x03);
+	mipi_dsi_usleep_range(dsi_ctx, 10000, 11000);
+	mipi_dsi_dcs_write_seq_multi(dsi_ctx, 0xb6, 0xa2);
+	mipi_dsi_usleep_range(dsi_ctx, 10000, 11000);
+	ea8074_test_key_off_lvl2(dsi_ctx);
+	mipi_dsi_usleep_range(dsi_ctx, 10000, 11000);
+
+	return dsi_ctx->accum_err;
+}
+
+static int ams621rc_init_sequence(struct mipi_dsi_multi_context *dsi_ctx)
+{
+	mipi_dsi_dcs_exit_sleep_mode_multi(dsi_ctx);
+	mipi_dsi_usleep_range(dsi_ctx, 10000, 11000);
+
+	mipi_dsi_dcs_set_page_address_multi(dsi_ctx, 0x0000, 0x0800 | 0x00c7);
+	mipi_dsi_usleep_range(dsi_ctx, 10000, 11000);
+
+	ea8074_test_key_on_lvl2(dsi_ctx);
+	mipi_dsi_dcs_write_seq_multi(dsi_ctx, 0xef, 0xf0, 0x31, 0x00, 0x33, 0x31, 0x14, 0x35);
+	mipi_dsi_dcs_write_seq_multi(dsi_ctx, 0xb0, 0x01);
+	mipi_dsi_dcs_write_seq_multi(dsi_ctx, 0xbb, 0x03);
+	mipi_dsi_dcs_write_seq_multi(dsi_ctx, 0xb0, 0x4f);
+	mipi_dsi_dcs_write_seq_multi(dsi_ctx, 0xcb, 0x00);
+	mipi_dsi_dcs_write_seq_multi(dsi_ctx, 0xb0, 0x6b);
+	mipi_dsi_dcs_write_seq_multi(dsi_ctx, 0xcb, 0x00);
+	mipi_dsi_dcs_write_seq_multi(dsi_ctx, 0xf7, 0x03);
+	mipi_dsi_dcs_write_seq_multi(dsi_ctx, 0xb0, 0x05);
+	mipi_dsi_dcs_write_seq_multi(dsi_ctx, 0xb1, 0x10);
+	mipi_dsi_dcs_write_seq_multi(dsi_ctx, 0xb0, 0x02);
+	mipi_dsi_dcs_write_seq_multi(dsi_ctx, 0xd5, 0x02, 0x17, 0x54, 0x14);
+	ea8074_test_key_off_lvl2(dsi_ctx);
+
+	ea8074_test_key_on_lvl2(dsi_ctx);
+	ea8074_test_key_on_lvl3(dsi_ctx);
+	mipi_dsi_dcs_write_seq_multi(dsi_ctx, 0xd2, 0x9f, 0xf0);
+	mipi_dsi_dcs_write_seq_multi(dsi_ctx, 0xb0, 0x0e);
+	mipi_dsi_dcs_write_seq_multi(dsi_ctx, 0xd2, 0x70);
+	mipi_dsi_dcs_write_seq_multi(dsi_ctx, 0xb0, 0x04);
+	mipi_dsi_dcs_write_seq_multi(dsi_ctx, 0xd2, 0x20);
+	ea8074_test_key_off_lvl2(dsi_ctx);
+	ea8074_test_key_off_lvl3(dsi_ctx);
+
+	mipi_dsi_dcs_set_tear_on_multi(dsi_ctx, MIPI_DSI_DCS_TEAR_MODE_VBLANK);
+
+	mipi_dsi_dcs_write_seq_multi(dsi_ctx, MIPI_DCS_WRITE_CONTROL_DISPLAY, 0x20);
+	mipi_dsi_dcs_write_seq_multi(dsi_ctx, MIPI_DCS_WRITE_POWER_SAVE, 0x00);
+	mipi_dsi_msleep(dsi_ctx, 110);
+
+	return dsi_ctx->accum_err;
+}
+
+static int samsung_ea8074_enable(struct drm_panel *panel)
+{
+	struct samsung_ea8074 *ctx = to_samsung_ea8074(panel);
+	struct mipi_dsi_multi_context dsi_ctx = { .dsi = ctx->dsi };
+
+	mipi_dsi_dcs_set_display_on_multi(&dsi_ctx);
+	mipi_dsi_msleep(&dsi_ctx, 120);
+
+	return dsi_ctx.accum_err;
+}
+
+static int samsung_ea8074_disable(struct drm_panel *panel)
+{
+	struct samsung_ea8074 *ctx = to_samsung_ea8074(panel);
+	struct mipi_dsi_multi_context dsi_ctx = { .dsi = ctx->dsi };
+
+	mipi_dsi_dcs_set_display_off_multi(&dsi_ctx);
+	mipi_dsi_usleep_range(&dsi_ctx, 17000, 18000);
+
+	mipi_dsi_dcs_enter_sleep_mode_multi(&dsi_ctx);
+	mipi_dsi_msleep(&dsi_ctx, 120);
+
+	return dsi_ctx.accum_err;
+}
+
+static int samsung_ea8074_prepare(struct drm_panel *panel)
+{
+	struct samsung_ea8074 *ctx = to_samsung_ea8074(panel);
+	struct mipi_dsi_multi_context dsi_ctx = { .dsi = ctx->dsi };
+	int ret;
+
+	ret = regulator_bulk_enable(ARRAY_SIZE(ea8074_supplies), ctx->supplies);
+	if (ret < 0)
+		return ret;
+
+	samsung_ea8074_reset(ctx);
+
+	ret = ctx->desc->init_sequence(&dsi_ctx);
+	if (ret < 0) {
+		gpiod_set_value_cansleep(ctx->reset_gpio, 1);
+		regulator_bulk_disable(ARRAY_SIZE(ea8074_supplies), ctx->supplies);
+		return ret;
+	}
+
+	return 0;
+}
+
+static int samsung_ea8074_unprepare(struct drm_panel *panel)
+{
+	struct samsung_ea8074 *ctx = to_samsung_ea8074(panel);
+
+	gpiod_set_value_cansleep(ctx->reset_gpio, 1);
+	regulator_bulk_disable(ARRAY_SIZE(ea8074_supplies), ctx->supplies);
+
+	return 0;
+}
+
+static int samsung_ea8074_get_modes(struct drm_panel *panel,
+					 struct drm_connector *connector)
+{
+	struct samsung_ea8074 *ctx = to_samsung_ea8074(panel);
+
+	return drm_connector_helper_get_modes_fixed(connector, ctx->desc->mode);
+}
+
+static const struct drm_panel_funcs samsung_ea8074_panel_funcs = {
+	.enable = samsung_ea8074_enable,
+	.prepare = samsung_ea8074_prepare,
+	.disable = samsung_ea8074_disable,
+	.unprepare = samsung_ea8074_unprepare,
+	.get_modes = samsung_ea8074_get_modes,
+};
+
+static int samsung_ea8074_bl_update_status(struct backlight_device *bl)
+{
+	struct mipi_dsi_device *dsi = bl_get_data(bl);
+	u16 brightness = backlight_get_brightness(bl);
+	int err;
+
+	dsi->mode_flags &= ~MIPI_DSI_MODE_LPM;
+
+	err = mipi_dsi_dcs_set_display_brightness_large(dsi, brightness);
+	if (err < 0)
+		return err;
+
+	dsi->mode_flags |= MIPI_DSI_MODE_LPM;
+
+	return 0;
+}
+
+static const struct backlight_ops samsung_ea8074_bl_ops = {
+	.update_status = samsung_ea8074_bl_update_status,
+};
+
+static struct backlight_device *
+samsung_ea8074_create_backlight(struct mipi_dsi_device *dsi)
+{
+	struct device *dev = &dsi->dev;
+	const struct backlight_properties props = {
+		.type = BACKLIGHT_RAW,
+		.brightness = 512,
+		.max_brightness = 1023,
+	};
+
+	return devm_backlight_device_register(dev, dev_name(dev), dev, dsi,
+					      &samsung_ea8074_bl_ops, &props);
+}
+
+static int samsung_ea8074_probe(struct mipi_dsi_device *dsi)
+{
+	struct device *dev = &dsi->dev;
+	struct samsung_ea8074 *ctx;
+	int ret;
+
+	ctx = devm_drm_panel_alloc(dev, struct samsung_ea8074, panel,
+				   &samsung_ea8074_panel_funcs,
+				   DRM_MODE_CONNECTOR_DSI);
+	if (IS_ERR(ctx))
+		return PTR_ERR(ctx);
+
+	ret = devm_regulator_bulk_get_const(dev,
+					    ARRAY_SIZE(ea8074_supplies),
+					    ea8074_supplies,
+					    &ctx->supplies);
+	if (ret < 0)
+		return dev_err_probe(dev, ret, "Failed to get regulators\n");
+
+	ctx->reset_gpio = devm_gpiod_get(dev, "reset", GPIOD_OUT_LOW);
+	if (IS_ERR(ctx->reset_gpio))
+		return dev_err_probe(dev, PTR_ERR(ctx->reset_gpio),
+				     "Failed to get reset-gpios\n");
+
+	mipi_dsi_set_drvdata(dsi, ctx);
+
+	ctx->dsi = dsi;
+	ctx->desc = of_device_get_match_data(dev);
+
+	dsi->lanes = 4;
+	dsi->format = MIPI_DSI_FMT_RGB888;
+	dsi->mode_flags = MIPI_DSI_CLOCK_NON_CONTINUOUS | MIPI_DSI_MODE_LPM;
+
+	ctx->panel.prepare_prev_first = true;
+
+	ctx->panel.backlight = samsung_ea8074_create_backlight(dsi);
+	if (IS_ERR(ctx->panel.backlight))
+		return dev_err_probe(dev, PTR_ERR(ctx->panel.backlight),
+				     "Failed to create backlight\n");
+
+	drm_panel_add(&ctx->panel);
+
+	ret = mipi_dsi_attach(dsi);
+	if (ret < 0) {
+		dev_err(dev, "Failed to attach to DSI host: %d\n", ret);
+		drm_panel_remove(&ctx->panel);
+		return ret;
+	}
+
+	return 0;
+}
+
+static void samsung_ea8074_remove(struct mipi_dsi_device *dsi)
+{
+	struct samsung_ea8074 *ctx = mipi_dsi_get_drvdata(dsi);
+	int ret;
+
+	ret = mipi_dsi_detach(dsi);
+	if (ret < 0)
+		dev_err(&dsi->dev, "Failed to detach from DSI host: %d\n", ret);
+
+	drm_panel_remove(&ctx->panel);
+}
+
+static const struct drm_display_mode l590nt06_mode = {
+	.clock = (1080 + 48 + 16 + 48) * (2160 + 20 + 12 + 28) * 60 / 1000,
+	.hdisplay = 1080,
+	.hsync_start = 1080 + 48,
+	.hsync_end = 1080 + 48 + 16,
+	.htotal = 1080 + 48 + 16 + 48,
+	.vdisplay = 2160,
+	.vsync_start = 2160 + 20,
+	.vsync_end = 2160 + 20 + 12,
+	.vtotal = 2160 + 20 + 12 + 28,
+	.width_mm = 68,
+	.height_mm = 137,
+	.type = DRM_MODE_TYPE_DRIVER | DRM_MODE_TYPE_PREFERRED,
+};
+
+static const struct drm_display_mode ams621rc_mode = {
+	.clock = (1080 + 56 + 18 + 56) * (2248 + 26 + 12 + 24) * 60 / 1000,
+	.hdisplay = 1080,
+	.hsync_start = 1080 + 56,
+	.hsync_end = 1080 + 56 + 18,
+	.htotal = 1080 + 56 + 18 + 56,
+	.vdisplay = 2248,
+	.vsync_start = 2248 + 26,
+	.vsync_end = 2248 + 26 + 12,
+	.vtotal = 2248 + 26 + 12 + 24,
+	.width_mm = 68,
+	.height_mm = 142,
+	.type = DRM_MODE_TYPE_DRIVER | DRM_MODE_TYPE_PREFERRED,
+};
+
+const struct samsung_ea8074_panel_desc l590nt06_display_desc = {
+	.mode = &l590nt06_mode,
+	.init_sequence = l590nt06_init_sequence,
+};
+
+const struct samsung_ea8074_panel_desc ams621rc_display_desc = {
+	.mode = &ams621rc_mode,
+	.init_sequence = ams621rc_init_sequence,
+};
+
+static const struct of_device_id samsung_ea8074_of_match[] = {
+	/* Comma three */
+	{ .compatible = "samsung,ea8074-l590nt06", .data = &l590nt06_display_desc },
+	 /* Huawei Mate 10 Pro */
+	{ .compatible = "samsung,ea8074-ams601nt01", .data = &l590nt06_display_desc },
+	/* Xiaomi Mi 8 */
+	{ .compatible = "samsung,ea8074-ams621rc", .data = &ams621rc_display_desc },
+	{ /* sentinel */ }
+};
+MODULE_DEVICE_TABLE(of, samsung_ea8074_of_match);
+
+static struct mipi_dsi_driver samsung_ea8074_driver = {
+	.probe = samsung_ea8074_probe,
+	.remove = samsung_ea8074_remove,
+	.driver = {
+		.name = "panel-samsung-ea8074",
+		.of_match_table = samsung_ea8074_of_match,
+	},
+};
+module_mipi_dsi_driver(samsung_ea8074_driver);
+
+MODULE_AUTHOR("David Heidelberg <david@ixit.cz>");
+MODULE_DESCRIPTION("DRM driver for the Samsung EA8074 panel");
+MODULE_LICENSE("GPL");
